pandajedi/jedidog/AtlasTaskWithholderWatchDog.py

```diff
@@ -44,230 +46,97 @@ def refresh(self):
-        # check for preassigned
-        if sitePreAssigned and preassignedSite not in scanSiteList:
-            scanSiteList = []
-            tmpLog.info('data is unavailable locally or remotely at preassigned site {0}'.format(preassignedSite))
-        elif len(scanSiteListOnDisk) > 0:
-            # use only disk sites
-            scanSiteList = list(scanSiteListOnDisk)
-        scanSiteList = self.get_pseudo_sites(scanSiteList, oldScanSiteList)
-        # dump
-        # for tmpSiteName in oldScanSiteList:
-        #     if tmpSiteName not in scanSiteList:
-        #         #tmpLog.info('  skip site={0} data is unavailable criteria=-input'.format(tmpSiteName))
-        #         pass
-        tmpLog.info('{0} candidates have input data'.format(len(scanSiteList)))
+    # get busy sites
+    def get_busy_sites(self, gshare):
+        busy_sites_list = []
+        for tmpPseudoSiteName in self.allSiteList:
+            tmpSiteSpec = self.siteMapper.getSite(tmpPseudoSiteName)
+            tmpSiteName = tmpSiteSpec.get_unified_name()
+            # get nQueue and nRunning
+            nRunning = AtlasBrokerUtils.getNumJobs(jobStatPrioMap, tmpSiteName, 'running', gshare)
+            nQueue = 0
+            for jobStatus in ['defined', 'assigned', 'activated', 'starting']:
+                nQueue += AtlasBrokerUtils.getNumJobs(jobStatPrioMap, tmpSiteName, jobStatus, gshare)
+            # busy sites
+            if (nQueue + 1)/(nRunning + 1) > 2:
+                busy_sites_list.append(tmpSiteName)
         # return
-        if not scanSiteList:
-            tmpLog.warning('no candidates')
-            return
-        return scanSiteList
+        return busy_sites_list
+
+    # # handle waiting jobs
+    # def do_make_tasks_pending(self, task_list):
+    #     tmpLog = MsgWrapper(logger, 'do_make_tasks_pending')
+    #     tmpLog.debug('start')
+    #     # check every x min
+    #     checkInterval = 20
+    #     # make task pending
+    #     for taskID, pending_reason in task_list:
+    #         tmpLog = MsgWrapper(logger, '< #ATM #KV do_make_tasks_pending jediTaskID={0}>'.format(taskID))
+    #         retVal = self.taskBufferIF.makeTaskPending_JEDI(taskID, reason=pending_reason)
+    #         tmpLog.debug('done with {0}'.format(retVal))
 
 
-    # handle waiting jobs
-    def do_make_tasks_pending(self, taskspec_list):
-        tmpLog = MsgWrapper(logger, 'do_make_tasks_pending')
-        tmpLog.debug('start')
-        # check every x min
-        checkInterval = 20
-        # make task pending
-        for taskspec, pending_reason in taskspec_list:
-            tmpLog = MsgWrapper(logger, '< #ATM #KV do_make_tasks_pending jediTaskID={0}>'.format(taskspec.jediTaskID))
-            retVal = self.taskBufferIF.makeTaskPending_JEDI(taskspec.jediTaskID, reason=pending_reason)
-            tmpLog.debug('done with {0}'.format(retVal))
+    # set tasks to be pending due to condition of data locality
+    def do_for_data_locality(self):
```

Comment:
I rewrote taskwithholder watchdog and here is the function about data locality check.
Is the algorithm okay?

And I have some questions:
-  How to get the list of prod_source_label, job_label, gshare , resource_type to loop over? Does it even matter whether to check them separately? (I can imagine check per gshare since nQ/nR is computed per gshare, but I don't quite understand why check per other  attributes especially resource_type)
- Which is the idea threshold of ioIntensity currentPriority should we set here?